src/Integration/State/StateManager.cs

```diff
@@ -123,11 +118,16 @@ public void SetProjects(ConnectionInformation connection, IEnumerable<SonarQubeP
             }
         }
 
-        public void SetBoundProject(SonarQubeProject project)
+        public void SetBoundProject(ConnectionInformation connection, SonarQubeProject project)
         {
             this.ClearBindingErrorNotifications();
-            ProjectViewModel projectViewModel = this.ManagedState.ConnectedServers.SelectMany(s => s.Projects).SingleOrDefault(p => p.Project == project);
+            
+            var serverViewModel = this.ManagedState.ConnectedServers.FirstOrDefault(s => s.Url == connection.ServerUri);
+            Debug.Assert(serverViewModel != null, "Expecting the connection to map to a single server");
+
+            var projectViewModel = serverViewModel?.Projects?.FirstOrDefault(p => SonarQubeProject.KeyComparer.Equals(p.Project.Key, project.Key));
```

Comment:
_ManagedState_ maintains a list of SonarQube servers, with each SonarQube server maintaining its own list of associated projects. Project keys are not guaranteed to be unique across SonarQube servers. The old code relied on matching on object identity to correctly match the supplied project to the correct server instance.

I've made this explicit; the caller setting the bound project is responsible for supplying the connection information too. This means the project lookup is now done using the project key, not on the project object identity.

Reply:
Update: I'll change this method to accept the server uri, organisation key and project keys directly.
That will make it more obvious that this method isn't holding on to the supplied parameters, but is using them to look up a project that should already exist.
Also, there is a bug in the new implementation - it doesn't take organisation into account when doing the lookup.